src/lib/replication/repl_dev/raft_state_machine.cpp

```diff
@@ -16,55 +16,22 @@ RaftStateMachine::RaftStateMachine(RaftReplDev& rd) : m_rd{rd} {
     m_success_ptr->put(0);
 }
 
-raft_buf_ptr_t RaftStateMachine::pre_commit_ext(nuraft::state_machine::ext_op_params const& params) {
-    // Leader precommit is processed in next callback, because this callback doesn't provide a way to stick a context
-    // which could contain the req structure in it.
-    if (!m_rd.is_leader()) {
-        int64_t lsn = s_cast< int64_t >(params.log_idx);
-        raft_buf_ptr_t data = params.data;
-
-        repl_req_ptr_t rreq = lsn_to_req(lsn);
-        RD_LOG(INFO, "Raft channel: Precommit rreq=[{}]", rreq->to_compact_string());
-        m_rd.m_listener->on_pre_commit(rreq->lsn, rreq->header, rreq->key, rreq);
-    }
-    return m_success_ptr;
-}
-
-void RaftStateMachine::after_precommit_in_leader(nuraft::raft_server::req_ext_cb_params const& params) {
-    repl_req_ptr_t rreq = repl_req_ptr_t(r_cast< repl_req_ctx* >(params.context));
-    link_lsn_to_req(rreq, int64_cast(params.log_idx));
-
-    RD_LOG(INFO, "Raft Channel: Proposed rreq=[{}]", rreq->to_compact_string());
-    m_rd.m_listener->on_pre_commit(rreq->lsn, rreq->header, rreq->key, rreq);
+static std::pair< sisl::blob, sisl::blob > header_only_extract(nuraft::buffer& buf) {
+    repl_journal_entry* jentry = r_cast< repl_journal_entry* >(buf.data_begin());
+    RELEASE_ASSERT_EQ(jentry->major_version, repl_journal_entry::JOURNAL_ENTRY_MAJOR,
+                      "Mismatched version of journal entry received from RAFT peer");
+    RELEASE_ASSERT_EQ(jentry->code, journal_type_t::HS_HEADER_ONLY,
+                      "Trying to extract header on non-header only entry");
+    sisl::blob const header = sisl::blob{uintptr_cast(jentry) + sizeof(repl_journal_entry), jentry->user_header_size};
+    sisl::blob const key = sisl::blob{header.cbytes() + header.size(), jentry->key_size};
+    return {header, key};
 }
 
-raft_buf_ptr_t RaftStateMachine::commit_ext(nuraft::state_machine::ext_op_params const& params) {
-    int64_t lsn = s_cast< int64_t >(params.log_idx);
-    raft_buf_ptr_t data = params.data;
-
-    repl_req_ptr_t rreq = lsn_to_req(lsn);
-    if (rreq == nullptr) { return m_success_ptr; }
-
-    RD_LOG(INFO, "Raft channel: Received Commit message rreq=[{}]", rreq->to_compact_string());
-    if (m_rd.is_leader()) {
-        // This is the time to ensure flushing of journal happens in leader
-        if (m_rd.m_data_journal->last_durable_index() < uint64_cast(lsn)) { m_rd.m_data_journal->flush(); }
-        rreq->state.fetch_or(uint32_cast(repl_req_state_t::LOG_FLUSHED));
-    }
-    if (rreq->state.load() & uint32_cast(repl_req_state_t::DATA_WRITTEN)) {
-        m_lsn_req_map.erase(rreq->lsn);
-        m_rd.report_committed(rreq);
-    }
-    return m_success_ptr;
-}
-
-uint64_t RaftStateMachine::last_commit_index() { return uint64_cast(m_rd.get_last_commit_lsn()); }
-
```

Comment:
Moved this precommit_ext and commit_ext to have linear flow in the source code, we first write into journal (and thus transform), then pre_commit and then commit.